File: 155.min-stack.0.go
/*
 * @lc app=leetcode id=155 lang=golang
 *
 * [155] Min Stack
 *
 * https://leetcode.com/problems/min-stack/description/
 *
 * algorithms
 * Easy (36.20%)
 * Total Accepted:    284.1K
 * Total Submissions: 782.4K
 * Testcase Example:  '["MinStack","push","push","push","getMin","pop","top","getMin"]\n[[],[-2],[0],[-3],[],[],[],[]]'
 *
 *
 * Design a stack that supports push, pop, top, and retrieving the minimum
 * element in constant time.
 *
 *
 * push(x) -- Push element x onto stack.
 *
 *
 * pop() -- Removes the element on top of the stack.
 *
 *
 * top() -- Get the top element.
 *
 *
 * getMin() -- Retrieve the minimum element in the stack.
 *
 *
 *
 *
 * Example:
 *
 * MinStack minStack = new MinStack();
 * minStack.push(-2);
 * minStack.push(0);
 * minStack.push(-3);
 * minStack.getMin();   --> Returns -3.
 * minStack.pop();
 * minStack.top();      --> Returns 0.
 * minStack.getMin();   --> Returns -2.
 *
 *
 */
type MinStack struct {
	Vals    []int
	MinVals []int
}

/** initialize your data structure here. */
func Constructor() MinStack {
	return MinStack{}
}

func (this *MinStack) Push(x int) {
	this.Vals = append(this.Vals, x)
	if len(this.MinVals) == 0 || x <= this.MinVals[len(this.MinVals)-1] {
		this.MinVals = append(this.MinVals, x)
	}
}

func (this *MinStack) Pop() {
	if len(this.Vals) <= 0 {
		return
	}

	if this.Vals[len(this.Vals)-1] == this.MinVals[len(this.MinVals)-1] {
		this.MinVals = this.MinVals[0 : len(this.MinVals)-1]
	}
	this.Vals = this.Vals[0 : len(this.Vals)-1]
}

func (this *MinStack) Top() int {
	return this.Vals[len(this.Vals)-1]
}

func (this *MinStack) GetMin() int {
	return this.MinVals[len(this.MinVals)-1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: 155.min-stack.0.rs
/*
 * @lc app=leetcode id=155 lang=rust
 *
 * [155] Min Stack
 */
struct MinStack {
    vec: Vec<i32>,
    min_vec: Vec<i32>,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {

    /** initialize your data structure here. */
    fn new() -> Self {
        MinStack{
            vec: Vec::new(),
            min_vec: Vec::new(),
        }
    }

    fn push(&mut self, x: i32) {
        self.vec.push(x);
        if self.min_vec.len() == 0 || self.get_min() >= x {
            self.min_vec.push(x);
        }
    }

    fn pop(&mut self) {
        if self.vec.pop().unwrap() == self.get_min() {
            self.min_vec.pop();
        }
    }

    fn top(&self) -> i32 {
        *self.vec.last().unwrap()
    }

    fn get_min(&self) -> i32 {
        *self.min_vec.last().unwrap()
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * let obj = MinStack::new();
 * obj.push(x);
 * obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: i32 = obj.get_min();
 */
